### ISSUE TITLE
Parallelize count_reads.py

### ISSUE THREAD

**Issue by canuse** [riboviz/riboviz#386]
The countReads processes (riboviz.count_reads.py) is running with only one core now, and is one of the most time consuming processes in the workflow. (For example, this process need 2h54m in Wallace dataset, while the  execution time of the whole process is ~5h under 16 cores total, 8 core each sample)

The core function of countReads is implemented in count_reads.py, function count_reads_df, which look like:
```python
def count_reads_df(config_file, input_dir, tmp_dir, output_dir):
    df = pd.DataFrame(columns=HEADER)
    rows = []
    rows.extend(input_fq(config_file, input_dir))
    rows.append(cutadapt_fq(tmp_dir))
    rows.extend(umi_tools_deplex_fq(tmp_dir))
    tmp_samples = [f.name for f in os.scandir(tmp_dir) if f.is_dir()]
    tmp_samples.sort()
    for sample in tmp_samples:
        rows.append(cutadapt_fq(tmp_dir, sample))
        rows.append(hisat2_fq(tmp_dir, sample, workflow_files.NON_RRNA_FQ,
                              "rRNA or other contaminating reads removed by alignment to rRNA index files"))
        rows.append(hisat2_sam(tmp_dir, sample, workflow_files.RRNA_MAP_SAM,
                               "Reads with rRNA and other contaminating reads removed by alignment to rRNA index files"))
        rows.append(hisat2_fq(tmp_dir, sample, workflow_files.UNALIGNED_FQ,
                              "Unaligned reads removed by alignment of remaining reads to ORFs index files"))
        rows.append(hisat2_sam(tmp_dir, sample, workflow_files.ORF_MAP_SAM,
                               "Reads aligned to ORFs index files"))
        rows.append(trim_5p_mismatch_sam(tmp_dir, sample))
        rows.append(umi_tools_dedup_bam(tmp_dir, output_dir, sample))
    rows = [row for row in rows if row is not None]
    df = df.append(rows)
    return df
```
We can see that the it runs different functions and add the result of each function into `rows`. This can be executed in parallel, by using the `multiprocessing` package, such as Pool and the apply_async to add tasks asynchronously.

I tested it on my VM and it is working, but when I tried to run a full size dataset on eddie, I got an OOM error, which may be caused by #385 .

**Comment by canuse** [comment]
I uses multiprocessing.Pool and apply_async to accelerate, and consider each whole function (i.e. hisat2_fq, hisat2_sam) as the target, so that the interface of that function do not need to change. The speedup and parallel efficiency is:
![image](https://user-images.githubusercontent.com/44812363/124740629-dfad1600-df12-11eb-9322-873ecfeeec5c.png)
![image](https://user-images.githubusercontent.com/44812363/124740640-e2a80680-df12-11eb-9abc-514a20ec972a.png)
Using 8 or 16 core, we can save about 10000 seconds, which is about a half time of the whole workflow.

The reason why parallel efficiency drops is that the some target function (like input_fq) needs ~2000-3000 seconds while some other like hisat2_sam only need several hundred or less seconds to execute. The unbalance makes the processes not balanced loaded. However, if we go into each function and parallel the loop inside, there will be two potential problems. The interface may be changed to pass multiprocessing.Pool settings and the feature that deamon multiprocess of python are not allowed to have child processes.

**Comment by mikej888** [comment]
Is the order of rows in the output file in the parallel version the same as for the serial version?

**Comment by canuse** [comment]
Yes. The multiprocessing.pool can guarantee the order of return value is the same order as the order of function calls, so I can control the order of rows in the output file to be the same.

**Comment by mikej888** [comment]
@canuse, thanks, that's good!

**Comment by canuse** [comment]
code updated in branch `parallel-countread-386` and tested locally

**Comment by canuse** [comment]
Like issue #376 , a smoke test may be needed. I run the Wallace_2020_JEC21 dataset on Eddie and it is working.
By the test, I want to double check that:

* If there are any errors that break the workflow
* If the results (actually only `read_counts.tsv` because other processes are not modified) are the same as current version

**Comment by mikej888** [comment]
Checks done on on Ubuntu 18.04, CentOS 7 2009, Ubuntu 20.04.2 (which vary in certain package versions) using branch `parallel-count-reads-386`, commit d1bffa2 Mon Aug 16 11:17:52 2021.

Integration tests:

```console
$ rm -rf vignette/output/ vignette/index vignette/tmp/ work/
$ nextflow run prep_riboviz.nf -ansi-log false -params-file vignette/vignette_config.yaml
$ pytest -vs riboviz/test/integration/test_integration.py --expected=../test-data-20210804 --skip-workflow
```

OK

All tests:

```console
$ pytest --ignore-glob="*integration*"
```

OK

**Comment by mikej888** [comment]
Other updates that need to be done either prior to a pull request being merged or after are as follows (if @canuse doesn't have time to do these then others in the team can, they're purely documentation-related updates):

* Update `riboviz/count_reads.py` with `pylint` and `pycodestyle` recommendations.

**Comment by canuse** [comment]
Output of `pycodestyle` and `pylint`:
```console
PS D:\GitHub\riboviz\riboviz> pycodestyle --statistics  .\count_reads.py
.\count_reads.py:38:80: E501 line too long (93 > 79 characters)
.\count_reads.py:42:80: E501 line too long (104 > 79 characters)
.\count_reads.py:43:80: E501 line too long (128 > 79 characters)
.\count_reads.py:44:80: E501 line too long (144 > 79 characters)
.\count_reads.py:106:80: E501 line too long (104 > 79 characters)
.\count_reads.py:107:80: E501 line too long (212 > 79 characters)
.\count_reads.py:113:80: E501 line too long (81 > 79 characters)
.\count_reads.py:115:80: E501 line too long (167 > 79 characters)
.\count_reads.py:145:80: E501 line too long (94 > 79 characters)
.\count_reads.py:153:80: E501 line too long (124 > 79 characters)
.\count_reads.py:160:80: E501 line too long (85 > 79 characters)
.\count_reads.py:421:80: E501 line too long (106 > 79 characters)
.\count_reads.py:501:80: E501 line too long (116 > 79 characters)
.\count_reads.py:504:80: E501 line too long (110 > 79 characters)
.\count_reads.py:506:80: E501 line too long (81 > 79 characters)
.\count_reads.py:507:80: E501 line too long (89 > 79 characters)
.\count_reads.py:511:80: E501 line too long (93 > 79 characters)
.\count_reads.py:512:80: E501 line too long (115 > 79 characters)
.\count_reads.py:513:80: E501 line too long (159 > 79 characters)
.\count_reads.py:514:80: E501 line too long (117 > 79 characters)
.\count_reads.py:515:80: E501 line too long (150 > 79 characters)
.\count_reads.py:516:80: E501 line too long (116 > 79 characters)
.\count_reads.py:517:80: E501 line too long (153 > 79 characters)
.\count_reads.py:518:80: E501 line too long (116 > 79 characters)
.\count_reads.py:519:80: E501 line too long (112 > 79 characters)
.\count_reads.py:520:80: E501 line too long (102 > 79 characters)
.\count_reads.py:521:80: E501 line too long (113 > 79 characters)
27      E501 line too long (93 > 79 characters)
PS D:\GitHub\riboviz\riboviz> pylint .\count_reads.py
************* Module count_reads
count_reads.py:42:0: C0301: Line too long (104/100) (line-too-long)
count_reads.py:43:0: C0301: Line too long (128/100) (line-too-long)
count_reads.py:44:0: C0301: Line too long (144/100) (line-too-long)
count_reads.py:106:0: C0301: Line too long (104/100) (line-too-long)
count_reads.py:107:0: C0301: Line too long (212/100) (line-too-long)
count_reads.py:115:0: C0301: Line too long (167/100) (line-too-long)
count_reads.py:153:0: C0301: Line too long (124/100) (line-too-long)
count_reads.py:421:0: C0301: Line too long (106/100) (line-too-long)
count_reads.py:501:0: C0301: Line too long (116/100) (line-too-long)
count_reads.py:504:0: C0301: Line too long (110/100) (line-too-long)
count_reads.py:512:0: C0301: Line too long (115/100) (line-too-long)
count_reads.py:513:0: C0301: Line too long (159/100) (line-too-long)
count_reads.py:514:0: C0301: Line too long (117/100) (line-too-long)
count_reads.py:515:0: C0301: Line too long (150/100) (line-too-long)
count_reads.py:516:0: C0301: Line too long (116/100) (line-too-long)
count_reads.py:517:0: C0301: Line too long (153/100) (line-too-long)
count_reads.py:518:0: C0301: Line too long (116/100) (line-too-long)
count_reads.py:519:0: C0301: Line too long (112/100) (line-too-long)
count_reads.py:520:0: C0301: Line too long (102/100) (line-too-long)
count_reads.py:521:0: C0301: Line too long (113/100) (line-too-long)
count_reads.py:118:35: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:146:15: W0703: Catching too general exception Exception (broad-except)
count_reads.py:146:8: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:159:11: W0703: Catching too general exception Exception (broad-except)
count_reads.py:159:4: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:201:11: W0703: Catching too general exception Exception (broad-except)
count_reads.py:201:4: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:210:0: R0914: Too many local variables (16/15) (too-many-locals)
count_reads.py:272:19: W0703: Catching too general exception Exception (broad-except)
count_reads.py:272:12: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:282:23: W0703: Catching too general exception Exception (broad-except)
count_reads.py:282:16: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:323:11: W0703: Catching too general exception Exception (broad-except)
count_reads.py:323:4: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:360:11: W0703: Catching too general exception Exception (broad-except)
count_reads.py:360:4: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:410:15: W0703: Catching too general exception Exception (broad-except)
count_reads.py:410:8: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:418:15: W0703: Catching too general exception Exception (broad-except)
count_reads.py:418:8: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:467:11: W0703: Catching too general exception Exception (broad-except)
count_reads.py:467:4: C0103: Variable name "e" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:499:4: C0103: Variable name "df" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:534:4: C0103: Variable name "df" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:604:13: C0103: Variable name "f1" doesn't conform to snake_case naming style (invalid-name)
count_reads.py:604:17: C0103: Variable name "f2" doesn't conform to snake_case naming style (invalid-name)

------------------------------------------------------------------
Your code has been rated at 8.13/10 (previous run: 8.13/10, +0.00)
```
The reason of `broad-except` and `invalid-name` is that I follow the naming style of other functions in this Python script, which uses `e` to catch an Broad Exception.

**Comment by mikej888** [comment]
The `E501 line too long` warnings for lines after line 100 and which occur within comment blocks can be fixed by introducing line breaks into the long sentences.

> The reason of broad-except and invalid-name is that I follow the naming style of other functions in this Python script, which uses e to catch an Broad Exception.

I'd agree that these warnings can be ignored.

**Comment by canuse** [comment]
It's ready to be merged into develop.

**Comment by mikej888** [comment]
After `update-file-names-411` has been merged into `develop`, I will merge `workflow-bug-fix-416`, `parallel-count-read-386`, `parallel-countread-386-workflow-bug-fix-416` into `develop`.

**Comment by mikej888** [comment]
Branched `parallel-countread-386-411-416` from `parallel-countread-386` and merged `workflow-bug-fix-416` (see #416 and pull request #419) and `parallel-countread-386-workflow-bug-fix-416 (now deleted). `parallel-countread-386-411-416` to be merged into `parallel-countread-386` and then `parallel-countread-386` pull request #424 to be merged into `develop`  post-merge of `workflow-bug-fix-416` pull request #419 (in turn dine after merge of `update-file-names-411` pull request #422).
